File: BMWCenter/Core/OBD/VehicleSnapshot.swift
import Foundation

struct VehicleSnapshot: Sendable, Equatable {
    var timestamp: Date = .now
    // basic
    var speedKmh: Double?
    var rpm: Double?
    var coolantC: Double?
    var fuelLevelPct: Double?
    var mafGs: Double?
    var mapKpa: Double?
    var intakeAirC: Double?
    var throttlePct: Double?
    var engineLoadPct: Double?
    var voltage: Double?
    var engineFuelRateLh: Double?
    var runtimeS: Double?
    // extended
    var oilTempC: Double?
    /// Automatic gearbox oil temperature (°C), BMW EGS Mode 22 DID 1E01.
    var transmissionOilTempC: Double?
    /// Engine oil pressure in bar (gauge), from BMW Mode 22 when available.
    var oilPressureBar: Double?
    /// BMW commanded boost setpoint (kPa), Mode 22 DID 2C11.
    var boostSetpointKpa: Double?
    /// BMW actual boost / charge pressure (kPa), Mode 22 DID 2C10.
    var boostActualKpa: Double?
    /// HPFP rail pressure in bar, Mode 22 DID 2B0D.
    var fuelRailBar: Double?
    /// Radiator outlet (°C), DID D3B2.
    var radiatorOutletC: Double?
    /// Intercooler air (°C), DID 2C0B.
    var intercoolerC: Double?
    /// BMW Mode 22 MAF in kg/h (DID 2C20).
    var mafKgh: Double?
    /// VANOS intake / exhaust cam (°), DID 2C30 / 2C31.
    var vanosIntakeDeg: Double?
    var vanosExhaustDeg: Double?
    /// Low-pressure fuel pump (bar), DID 2B02.
    var lowPressureFuelBar: Double?
    /// Alternator voltage (V), DID 2D01.
    var alternatorVoltage: Double?
    /// Battery state of charge (%), DID 2D04.
    var batterySocPct: Double?
    var baroKpa: Double?
    var ambientC: Double?
    var stftBank1: Double?
    var ltftBank1: Double?
    var stftBank2: Double?
    var ltftBank2: Double?
    var catalystC: Double?
    var fuelRailKpa: Double?
    var pedalPct: Double?
    var timingAdvance: Double?
    var distanceSinceClearKm: Double?
    var distanceWithMILKm: Double?

    var isEngineRunning: Bool { (rpm ?? 0) > 300 }
    var isStale: Bool { Date.now.timeIntervalSince(timestamp) > 3.0 }

    /// Turbo boost (relative). Prefers BMW Mode 22 charge pressure; else MAP − baro.
    var boostKpa: Double? {
        if let absolute = boostActualKpa {
            let baro = baroKpa ?? 101.325
            return absolute - baro
        }
        guard let m = mapKpa, let b = baroKpa else { return nil }
        return m - b
    }

    var boostBar: Double? { boostKpa.map { $0 / 100.0 } }

    /// Absolute charge pressure from BMW DID 2C10 (bar), when present.
    var boostActualBar: Double? { boostActualKpa.map { $0 / 100.0 } }
}

struct DTC: Sendable, Equatable, Identifiable, Hashable {
    enum Status: String, Sendable {
        case stored, pending, permanent
    }

    var id: String { code }
    let code: String
    var status: Status
    var descriptionKey: String?
    /// Short plain-language meaning from DTCCatalog (localized).
    var summary: String?
    var severity: String?
    var firstSeen: Date

    init(
        code: String,
        status: Status = .stored,
        descriptionKey: String? = nil,
        summary: String? = nil,
        severity: String? = nil,
        firstSeen: Date = .now
    ) {
        self.code = code
        self.status = status
        self.descriptionKey = descriptionKey
        self.summary = summary
        self.severity = severity
        self.firstSeen = firstSeen
    }
}

File: mobile/src/core/obd/vehicleSnapshot.ts
export interface VehicleSnapshot {
  timestamp: number;
  speedKmh?: number;
  rpm?: number;
  coolantC?: number;
  fuelLevelPct?: number;
  mafGs?: number;
  mapKpa?: number;
  intakeAirC?: number;
  throttlePct?: number;
  engineLoadPct?: number;
  voltage?: number;
  engineFuelRateLh?: number;
  runtimeS?: number;
  oilTempC?: number;
  transmissionOilTempC?: number;
  oilPressureBar?: number;
  boostSetpointKpa?: number;
  boostActualKpa?: number;
  fuelRailBar?: number;
  radiatorOutletC?: number;
  intercoolerC?: number;
  mafKgh?: number;
  vanosIntakeDeg?: number;
  vanosExhaustDeg?: number;
  lowPressureFuelBar?: number;
  alternatorVoltage?: number;
  batterySocPct?: number;
  baroKpa?: number;
  ambientC?: number;
  stftBank1?: number;
  ltftBank1?: number;
  stftBank2?: number;
  ltftBank2?: number;
  catalystC?: number;
  fuelRailKpa?: number;
  pedalPct?: number;
  timingAdvance?: number;
  distanceSinceClearKm?: number;
  distanceWithMILKm?: number;
}

export function emptySnapshot(): VehicleSnapshot {
  return { timestamp: Date.now() };
}

export function isEngineRunning(s: VehicleSnapshot): boolean {
  return (s.rpm ?? 0) > 300;
}

export function isStale(s: VehicleSnapshot): boolean {
  return Date.now() - s.timestamp > 3000;
}

/** Turbo boost (relative kPa). Prefers BMW Mode 22 charge pressure; else MAP - baro. */
export function boostKpa(s: VehicleSnapshot): number | undefined {
  if (s.boostActualKpa != null) {
    const baro = s.baroKpa ?? 101.325;
    return s.boostActualKpa - baro;
  }
  if (s.mapKpa == null || s.baroKpa == null) return undefined;
  return s.mapKpa - s.baroKpa;
}

export function boostBar(s: VehicleSnapshot): number | undefined {
  const kpa = boostKpa(s);
  return kpa == null ? undefined : kpa / 100;
}

export function boostActualBar(s: VehicleSnapshot): number | undefined {
  return s.boostActualKpa == null ? undefined : s.boostActualKpa / 100;
}

export type DTCStatus = "stored" | "pending" | "permanent";

export interface DTC {
  code: string;
  status: DTCStatus;
  descriptionKey?: string;
  summary?: string;
  severity?: string;
  firstSeen: number;
}
